### ISSUE TITLE
Incorect results on big data set(11ml records)

### ISSUE THREAD

**Issue by DavidRimskiy** [nmslib/nmslib#486]
Hello,
I'm trying to run cosinesimil_sparse space with hnsw methods on 11 millions records dataset.
And when i try to fuzzy match company name i have 60/40 results, i 100% confident that vector consist company.
Then i create index, and run knnQueryBatch by k=30, and in 60 percent times in top 30 isn't needed company(company names are not similar at all)
in 40% times algorithm find correct company.
My create index parameters: are {  
        "M": 16,
        "indexThreadQty": 24,
        "efConstruction": 100,
        "post": 0,
        }
i create tfidf by TfidfVectorizer from sklearn.
i try m = 16, efConstruction = 512 - bad results
m = 20, efConstruction = 100,200 - bad results
m = 100, efConstruction = 100,200 - bad results(
Thanks for reply)

**Comment by searchivarius** [comment]
Hi @DavidRimskiy this is not an error per se, there may be multiple reasons why it doesn't perform well:
1. possible duplicates in the data set
2. low value of the parameter `ef`.
As a sanity check you can try using an inverted index (https://github.com/nmslib/nmslib/blob/master/manual/methods.md) : https://github.com/nmslib/nmslib/blob/master/manual/methods.md

It can be slowish, but results will be exact.

**Comment by DavidRimskiy** [comment]
> Hi @DavidRimskiy this is not an error per se, there may be multiple reasons why it doesn't perform well:
> 
> 1. possible duplicates in the data set
> 2. low value of the parameter `ef`.
>    As a sanity check you can try using an inverted index (https://github.com/nmslib/nmslib/blob/master/manual/methods.md) : https://github.com/nmslib/nmslib/blob/master/manual/methods.md
> 
> It can be slowish, but results will be exact.

Thanks for answear)
I've tried before inverted index (space:negdotprod_sparse_fast method:simple_ividix)
but without custom create index parameters and results was bad.
Vector did not consist duplicates, and entire dataframe too, i drop duplicates each time before run script.
By your answear i need to low efConstruction to 50 or lower, or replace it by efSearch = 50?

**Comment by searchivarius** [comment]
Does bad mean slow or inaccurate? Normally (modulo some possible bugs) it should be exact matching. Don't forget that the highest cosine similarity doesn't necessarily mean you get a result you want. In particular, `negdotprod_sparse_fast` is the maximum inner product search. So the highest result don't indicate the best match per se: you can have entries with multiple repeating words and they will have a higher score. Whether results are `good` or `bad` first needs to be compared with the brute-force search. If HNSW has a similar recall, it's the problem with encoding and the similarity metric rather than with k-NN search per se.

With respect to parameter selection, reasonable values are M=30, efConstruction=200, efSearch in the range of 50 to 2000. Oftentimes, it needs to be pretty high.

**Comment by DavidRimskiy** [comment]
> Does bad mean slow or inaccurate? Normally (modulo some possible bugs) for the inverted index it should be exact matching.
> 
> Don't forget that the highest similarity doesn't necessarily mean you get a result you want. In particular, `negdotprod_sparse_fast` is the maximum inner product search. So the highest result don't indicate the best match per se: you can have entries with multiple repeating words and they will have a higher score. Whether results are `good` or `bad` first needs to be compared with the brute-force search. If HNSW has a similar recall, it's the problem with encoding and the similarity metric rather than with k-NN search per se.
> 
> With respect to parameter selection, reasonable values are M=30, efConstruction=200, efSearch in the range of 50 to 2000. Oftentimes, it needs to be pretty high.

Thanks for Answer)
bad mean inaccurate)
I will try with your suggestions and let you know.

**Comment by DavidRimskiy** [comment]
> Does bad mean slow or inaccurate? Normally (modulo some possible bugs) for the inverted index it should be exact matching.
> 
> Don't forget that the highest similarity doesn't necessarily mean you get a result you want. In particular, `negdotprod_sparse_fast` is the maximum inner product search. So the highest result don't indicate the best match per se: you can have entries with multiple repeating words and they will have a higher score. Whether results are `good` or `bad` first needs to be compared with the brute-force search. If HNSW has a similar recall, it's the problem with encoding and the similarity metric rather than with k-NN search per se.
> 
> With respect to parameter selection, reasonable values are M=30, efConstruction=200, efSearch in the range of 50 to 2000. Oftentimes, it needs to be pretty high.

Hi, i tried your suggest approach, and the result on 11ml records
cosinesimil_sparse space with hnsw method:
{M = 30, efConstruction = 200, ef = 2000} = inaccurate
{M = 30, efConstruction = 200, ef = 1000} = inaccurate
{M = 20, efConstruction = 200, ef = 200} = inaccurate
negdotprod_sparse_fast = inacurate(
but i tried cosinesimil_sparse space with hnsw method locally 1Ram 4cpu on 1 mln record
with {M = 16, efConstruction = 100, ef = 100} without ef and result was inaccurate, but with ef = 100 result was perfect)

**Comment by searchivarius** [comment]
Great!
